### PR TITLE
[BEAM-801, BEAM-787] Improvements to Splittable DoFn

### PR DESCRIPTION
**Pull request by jkff**
Makes Splittable DoFn be more like a real DoFn:
- Adds support for side inputs and outputs to SDF
- Teaches `ProcessFn` to work with exploded windows inside the
  `KeyedWorkItem`. It works with them by un-exploding the windows
  in the `Iterable<WindowedValue<ElementAndRestriction>>` into a
  single `WindowedValue`, since the values and timestamps are
  guaranteed to be the same.

Makes SplittableParDo.ProcessFn not use the (now unavailable)
OldDoFn state and timers API:
- Makes `ProcessFn` be a primitive transform with its own
  `ParDoEvaluator`. As a nice side effect, this enables the runner to
  provide additional hooks into it - e.g. for giving the runner access
  to the restriction tracker (in later PRs)
- For consistency, moves declaration of `GBKIntoKeyedWorkItems`
  primitive transform into `SplittableParDo`, alongside the
  `SplittableProcessElements` transform
- Preserves compressed representation of `WindowedValue`'s in
  `PushbackSideInputDoFnRunner`

TODO: Tests for new features
TODO: Trigger on the GBK between initial split and ProcessFn
should be like the ReshuffleTrigger.
R: @kennknowles @tgroh

### CODE REVIEW

File: runners/core-java/src/main/java/org/apache/beam/runners/core/PushbackSideInputDoFnRunner.java [hunk 1]
```diff
@@ -74,17 +74,29 @@ public void startBundle() {
...
+        pushedBackWindowsBuilder.add(mainInputWindow);
       }
     }
-    return pushedBack.build();
+    ImmutableList<BoundedWindow> readyWindows = readyWindowsBuilder.build();
+    ImmutableList<BoundedWindow> pushedBackWindows = pushedBackWindowsBuilder.build();
+    if (!readyWindows.isEmpty()) {
+      processElement(
+          WindowedValue.of(
+              elem.getValue(), elem.getTimestamp(), readyWindows, elem.getPane()));
```
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/PushbackSideInputDoFnRunner.java]: I like this. Can you update the javadoc on `SimpleDoFnRunner` to make it clear that it owns exploding things when the value observes the window? I've felt that it was somewhat unclear before now.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java [hunk 2]
```diff
@@ -78,27 +90,92 @@
...
  */
 @Experimental(Experimental.Kind.SPLITTABLE_DO_FN)
-public class SplittableParDo<
-        InputT, OutputT, RestrictionT, TrackerT extends RestrictionTracker<RestrictionT>>
-    extends PTransform<PCollection<InputT>, PCollection<OutputT>> {
-  private final DoFn<InputT, OutputT> fn;
-  private final DoFnSignature signature;
+public class SplittableParDo {
+  /** The version of {@link SplittableParDo} for a {@link ParDo} with a single main output. */
+  public static class Bound<InputT, OutputT, RestrictionT>
```
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java]: [BEAM-447](https://issues.apache.org/jira/browse/BEAM-447) "stop referring to types as `Bound`/`Unbound`".

Especially here where there is no such duality.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java [hunk 3]
```diff
@@ -225,49 +376,73 @@ public void processElement(ProcessContext context) {
...
-      elementTag =
+      this.elementTag =
           StateTags.value("element", WindowedValue.getFullCoder(elementCoder, this.windowCoder));
-      DoFnInvoker<InputT, OutputT> invoker = DoFnInvokers.INSTANCE.newByteBuddyInvoker(fn);
-      restrictionTag = StateTags.value("restriction", restrictionCoder);
+      this.restrictionTag = StateTags.value("restriction", restrictionCoder);
     }
 
-    @Override
+    public void setStateInternalsFactory(StateInternalsFactory<String> stateInternalsFactory) {
```
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java]: Should be able to pass these to the constructor. That's actually almost the whole reason `StateInternalsFactory` is a factory.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java [hunk 4]
```diff
@@ -225,49 +376,73 @@ public void processElement(ProcessContext context) {
...
+      this.stateInternalsFactory = stateInternalsFactory;
+    }
+
+    public void setTimerInternalsFactory(TimerInternalsFactory<String> timerInternalsFactory) {
+      this.timerInternalsFactory = timerInternalsFactory;
+    }
+
+    public void setOutputManager(DoFnRunners.OutputManager outputManager) {
```
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java]: A `DoFn` should not depend on `DoFnRunner` and definitely shouldn't have a handle on `DoFnRunners.OutputManager`. A `DoFnRunner` is just one way of running a `DoFn` that we provide as a convenience.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java [hunk 5]
```diff
@@ -225,49 +376,73 @@ public void processElement(ProcessContext context) {
...
       WatermarkHoldState<GlobalWindow> holdState =
-          c.windowingInternals().stateInternals().state(stateNamespace, watermarkHoldTag);
+          stateInternals.state(stateNamespace, watermarkHoldTag);
 
       ElementAndRestriction<WindowedValue<InputT>, RestrictionT> elementAndRestriction;
       if (isSeedCall) {
         // The element and restriction are available in c.element().
         WindowedValue<ElementAndRestriction<InputT, RestrictionT>> windowedValue =
-            Iterables.getOnlyElement(c.element().elementsIterable());
+            implodeWindows(c.element().elementsIterable());
```
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java]: In addition to explain that you are allowed to do this, please explain why you do it (to protect against future edits).

File: runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java [hunk 6]
```diff
@@ -339,17 +540,17 @@ public PaneInfo pane() {
         }
 
         public void output(OutputT output) {
-          baseContext
-              .windowingInternals()
-              .outputWindowedValue(
-                  output, element.getTimestamp(), element.getWindows(), element.getPane());
+          outputManager.output(
```
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java]: To remove the dependence on `OutputManager`, here are ideas:

 - Can you make the output type simply be `WindowedValue` (with appropriate compensation afterwards). That would be an accurate description of the `DoFn` anyhow.
 - If you need to output to a particular window, `ReduceFnRunner` also needs a one-method interface with `outputWindowedValue(...)` to remove its dependence on `WindowingInternals` so you could create and use it here.
 - If you need to output a compressed representation, then you'll need a new interface different from the above, as `ReduceFnRunner` only outputs to one window at a time.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java [hunk 7]
```diff
@@ -369,17 +570,17 @@ public PipelineOptions getPipelineOptions() {
...
+                  output, element.getTimestamp(), element.getWindows(), element.getPane()));
+          noteOutput();
         }
 
         public <T> void sideOutputWithTimestamp(TupleTag<T> tag, T output, Instant timestamp) {
-          // TODO: I'm not sure how to implement this correctly: there's no
-          // "internals.sideOutputWindowedValue".
-          throw new UnsupportedOperationException(
-              "Side outputs not yet supported by splittable DoFn");
+          outputManager.output(
```
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java]: There should be a `sideOutputWindowedValue` to match `outputWindowedValue`, for sure, on whatever we end up with.

File: runners/core-java/src/test/java/org/apache/beam/runners/core/PushbackSideInputDoFnRunnerTest.java [hunk 8]
```diff
@@ -162,9 +161,14 @@ public void processElementSideInputNotReadySomeWindows() {
...
-    assertThat(underlying.inputElems,
-        containsInAnyOrder(WindowedValue.of(2, new Instant(-2), littleWindow, PaneInfo.NO_FIRING),
-            WindowedValue.of(2, new Instant(-2), bigWindow, PaneInfo.NO_FIRING)));
+    assertThat(
+        underlying.inputElems,
+        containsInAnyOrder(
+            WindowedValue.of(
+                2,
+                new Instant(-2),
+                ImmutableList.of(littleWindow, bigWindow),
```
Comment (kennknowles) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/PushbackSideInputDoFnRunnerTest.java]: I am OK with, but not excited about, the idea of making maintenance of the compressed representation part of the spec. In general I'd like to be as liberal as possible about allowing different areas of the SDK and runners-core choose between alternative equivalent representations.

File: runners/core-java/src/test/java/org/apache/beam/runners/core/SplittableParDoTest.java [hunk 9]
```diff
@@ -146,18 +151,23 @@ public void testBoundednessForUnboundedFn() {
...
         makeBoundedCollection(pipeline)
-            .apply("unbounded to bounded", new SplittableParDo<>(unboundedFn))
+            .apply("unbounded to bounded", new SplittableParDo.Bound<>(ParDo.of(unboundedFn)))
             .isBounded());
     assertEquals(
         "Applying an unbounded SDF to an unbounded collection produces an unbounded collection",
         PCollection.IsBounded.UNBOUNDED,
         makeUnboundedCollection(pipeline)
-            .apply("unbounded to unbounded", new SplittableParDo<>(unboundedFn))
+            .apply("unbounded to unbounded", new SplittableParDo.Bound<>(ParDo.of(unboundedFn)))
```
Comment (kennknowles) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/SplittableParDoTest.java]: Referring to my prior comments, all the things `set` just below should be passed to this constructor.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/state/TimerInternalsFactory.java [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+import org.apache.beam.sdk.util.TimerInternals;
+
+/**
+ * A factory for providing {@link TimerInternals} for a particular key.
+ *
+ * <p>Because it will generally be embedded in a {@link org.apache.beam.sdk.transforms.DoFn DoFn},
+ * albeit at execution time, it is marked {@link Serializable}.
+ */
+@Experimental(Kind.STATE)
+public interface TimerInternalsFactory<K> {
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/state/TimerInternalsFactory.java]: +1 we need this

**Comment by jkff** [comment]
This depends on #1353 - look at the two other commits.
Also, the original title of the PR is not quite as entertaining anymore :(
Still working on tests.

File: runners/core-java/src/test/java/org/apache/beam/runners/core/SplittableParDoTest.java (hunk 9 above)
Comment (jkff) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/SplittableParDoTest.java]: Unfortunately that's not possible - there's a cyclic dependency between DoFnTester and ProcessFn: DoFnTester is constructed with ProcessFn, but ProcessFn needs an OutputWindowedValue that will refer to DoFnTester.

Well, it's possible, with some trickery, but it seems not worth it. I think setters are a fine way of wiring together objects with a cyclic dependency.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java (hunk 6 above)
Comment (jkff) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java]: Done in #1353.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java (hunk 5 above)
Comment (jkff) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java]: Done.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java (hunk 4 above)
Comment (jkff) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java]: Done with the help of #1353 .

File: runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java (hunk 3 above)
Comment (jkff) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java]: See comment below.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java (hunk 2 above)
Comment (jkff) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java]: Done.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/PushbackSideInputDoFnRunner.java (hunk 1 above)
Comment (jkff) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/PushbackSideInputDoFnRunner.java]: Done.

File: runners/core-java/src/test/java/org/apache/beam/runners/core/SplittableParDoTest.java (hunk 9 above)
Comment (kennknowles) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/SplittableParDoTest.java]: This is quite an interesting situation. `DoFnTester` is for users, and it plays the role of the runner/worker in this regard.

So for a user who actually wants to test their splittable `DoFn`, the conversion to use any of the technology you have included should occur within `DoFnTester`, privately. But, of course, `DoFnTester` cannot access `runners-core`. I am not familiar enough with the preceding work on splittable `DoFn` to know how this should be resolved.

Often (as you can see from the internals in `DoFnTester`) we end up needing a slight variant on the details of what would otherwise come from `runners-core`. So in this case having a near-duplicate implementation there is not so bad. In the case of a splittable `DoFn` I would expect that the `DoFnTester` code to exercise splittability would diverge even further from this particular implementation in terms of `StateInternals`, etc.

If we stick to hacking on `setXYZ` methods, I would prefer to add `DoFnTester.getOutputWindowedValue(...)` accessor and `ProcessFn.setOutputWindowedValue(...)`, or appropriate variations. This will limit the use of this pattern to just testing, while all worker-side uses will be able to pass things through the constructor without a mutator.

**Comment by kennknowles** [comment]
I've left some comments on particular commits, but between conflicts and the underlying PR, just to clear up my tooling, I am going to defer reviewer assignment for now.

R: -@kennknowles

File: runners/core-java/src/test/java/org/apache/beam/runners/core/SplittableParDoTest.java (hunk 9 above)
Comment (jkff) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/SplittableParDoTest.java]: Note that in this test, I'm testing ProcessFn, rather than a user's splittable DoFn. Agreed that I'll need to make DoFnTester usable with user splittable DoFn's at some point.

Unfortunately I can't have DoFnTester.getOutputWindowedValue, because OutputWindowedValue is declared in runners-core and I'd like to keep it there. It is not a useful part of the SDK; like ProcessFn, it should be part of runners-core.

**Comment by jkff** [comment]
Fixed the various build warnings and a test error, PTAL.

**Comment by jkff** [comment]
The remaining test error is fixed by https://github.com/apache/incubator-beam/pull/1367

**Comment by jkff** [comment]
Rebased onto recent changes merging single- and multi-output ParDo evaluator factories. SplittableParDo now overrides only the multi-output ParDo (because single-output is overridden into multi-output).

However, I had to restore TransformHooks in ParDoEvaluatorFactory, because splittable and non-splittable are treated differently. But at least I could remove one type argument (POutput type - which is now always PCollectionTuple).

**Comment by jkff** [comment]
...Aaaand while addressing your comment, I realized that things are a little bit more complex, and SDF is failing to invoke the actual fn's Setup/Teardown/StartBundle/FinishBundle methods. Oh well. I'll work on this after Thanksgiving.

**Comment by jkff** [comment]
OK, should be all good. PTAL!

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/SplittableProcessElementsEvaluatorFactory.java [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+import org.joda.time.Instant;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class SplittableProcessElementsEvaluatorFactory<InputT, OutputT, RestrictionT>
+    implements TransformEvaluatorFactory {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(SplittableProcessElementsEvaluatorFactory.class);
+
+  private final LoadingCache<DoFn<?, ?>, DoFnLifecycleManager> fnClones;
```
Comment (kennknowles) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/SplittableProcessElementsEvaluatorFactory.java]: Can you really not re-use the logic in `ParDoEvaluatorFactory` rather than copy and paste?

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/SplittableProcessElementsEvaluatorFactory.java [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+                  stepContext,
+                  application,
+                  application.getInput().getWindowingStrategy(),
+                  processFn,
+                  transform.getSideInputs(),
+                  transform.getMainOutputTag(),
+                  transform.getSideOutputTags().getAll(),
+                  application.getOutput().getAll());
+
+      processFn.setStateInternalsFactory(
```
Comment (kennknowles) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/SplittableProcessElementsEvaluatorFactory.java]: I understand that you need to have `setStateInternalsFactory` in order to use this with `DoFnTester`. But all non-test code should follow the normal pattern of passing dependencies to the constructor.

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/SplittableDoFnTest.java [hunk 13]
```diff
@@ -228,4 +236,172 @@ public void testPairWithIndexWindowedTimestamped() throws ClassNotFoundException
...
+    }
+
+    @Teardown
+    public void tearDown() {
+      assertEquals(State.OUTSIDE_BUNDLE, state);
+      state = State.TORN_DOWN;
+    }
+  }
+
+  @Test
```
Comment (kennknowles) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/SplittableDoFnTest.java]: This should be `@NeedsRunner`

**Comment by kennknowles** [comment]
R: @kennknowles (since I removed before)

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/SplittableProcessElementsEvaluatorFactory.java (hunk 12 above)
Comment (jkff) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/SplittableProcessElementsEvaluatorFactory.java]: If I'm using ParDoEvaluator to evaluate processFn at all, I'm not sure how I can pass stateInternalsFactory, timerInternalsFactory and outputWindowedValue to the constructor - they are different in every bundle. Do you have suggestions?

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/SplittableDoFnTest.java (hunk 13 above)
Comment (jkff) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/SplittableDoFnTest.java]: This whole test class should be, but I'd rather do that in a separate PR (move it to runners-core and exclude it from all runners but the direct one, like you did with Stateful ParDo tests).

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/SplittableProcessElementsEvaluatorFactory.java (hunk 11 above)
Comment (jkff) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/SplittableProcessElementsEvaluatorFactory.java]: Done; moved a few things around and now the whole ProcessFn is cacheable.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/SplittableProcessElementsEvaluatorFactory.java (hunk 12 above)
Comment (kennknowles) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/SplittableProcessElementsEvaluatorFactory.java]: The evaluator, such as `ParDoEvaluator`, is one per bundle. So `SplittableProcessElementsEvaluatorFactory.createEvaluator` is safe to pass them right here.

**Comment by jkff** [comment]
@kennknowles says he'd like to pass the review to @tgroh (with some followup offline comments), so
R: @tgroh

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/SplittableProcessElementsEvaluatorFactory.java [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+        parDoEvaluator =
+            delegateFactory.createParDoEvaluator(
+                application,
+                transform.getSideInputs(),
+                transform.getMainOutputTag(),
+                transform.getSideOutputTags().getAll(),
+                stepContext,
+                fnManager);
+
+    processFn.setStateInternalsFactory(
```
Comment (tgroh) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/SplittableProcessElementsEvaluatorFactory.java]: The `ProcessFn` is to all appearances a System DoFn, as it requires intervention from the system to be properly executed. As such, you should construct it on demand, as it's an implementation detail of `SplittableDoFn` - which means you can construct a new instance within the evaluator for each bundle. The `SplittableParDo` should instead carry as its payload the `SplittableDoFn`, and the `ProcessFn` can be constructed with appropriate `StateInternals`, `TimerInternals`, and `OutputWindowedValue` (`WindowedOutputReciever`?). This is almost the same pattern we formerly used for `GroupAlsoByWindowsDoFn` before that code was entirely migrated to be entirely a primitive in https://github.com/apache/incubator-beam/pull/1251

You can still invoke the methods via `ParDoEvaluator` (though I think generally that's probably discouraged, as the runner will probably want to interact with a `SplittableDoFn` in ways that can't easily be expressed in terms of a `ParDo`).

Comment (tgroh) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/SplittableProcessElementsEvaluatorFactory.java]: Moving to `runners-core` also gels with reasoning behind the `GroupByKeyViaGroupAlsoByWindow` expansion being part of `runners-core` - the user has expressed a splittable `DoFn`, and the runner can use the default implementation with minimal wiring by going through an SDK-provided expansion.

Comment (jkff) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/SplittableProcessElementsEvaluatorFactory.java]: OK, I changed the code to create a ProcessFn at evaluation time. However, I still can't pass StateInternals, TimerInternals and OutputWindowedValue in the constructor due to cyclic dependencies between these and the surrounding objects - e.g. DoFnTester (e.g. the ProcessFn depends on DoFnTester.getTimerInternals(), and DoFnTester obviously depends on the ProcessFn) and ParDoEvaluator (similar cyclic dependency with OutputWindowedValue).

I discussed this at length with @kennknowles and we decided to keep these setters and give up on moving this setup into the ProcessFn's constructor because there's no easy way to do this and because the set-up procedure of a ProcessFn is going to change in nearby PRs anyway (where I'll be adding more runner hooks). But the current PR, I think, is an acceptable intermediate state.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java [hunk 15]
```diff
@@ -508,6 +517,20 @@ public void onTimer(TimerInternals.TimerData timer) throws Exception {
     return MoreObjects.firstNonNull(elems, Collections.<WindowedValue<T>>emptyList());
   }
 
+  @SuppressWarnings({"unchecked", "rawtypes"})
+  public <T> List<WindowedValue<T>> getOrCreateOutput(TupleTag<T> tag) {
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java]: Unless there's a mirrored `getOuput` which throws if it doesn't exist, this should just be `getOutput`

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoEvaluatorFactory.java [hunk 16]
```diff
@@ -134,4 +153,8 @@ public void cleanup() throws Exception {
       throw e;
     }
   }
+
+  public DoFnLifecycleManager getFnClone(DoFn<?, ?> fn) {
```
Comment (jkff) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoEvaluatorFactory.java]: Renamed.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java (hunk 15 above)
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java]: Renamed this one and getOutput to respectively getMutableOutput and getImmutableOutput, and fixed a couple more small things in DoFnTester that followed.
